rechunker/executors/pipeline.py

```diff
@@ -0,0 +1,82 @@
+import itertools
+import math
+from typing import Iterable, Iterator, Tuple, TypeVar
+
+import dask
+import numpy as np
+
+from rechunker.types import (
+    CopySpec,
+    MultiStagePipeline,
+    ParallelPipelines,
+    ReadableArray,
+    Stage,
+    WriteableArray,
+)
+
+
+def chunk_keys(
+    shape: Tuple[int, ...], chunks: Tuple[int, ...]
+) -> Iterator[Tuple[slice, ...]]:
+    """Iterator over array indexing keys of the desired chunk sized.
+
+    The union of all keys indexes every element of an array of shape ``shape``
+    exactly once. Each array resulting from indexing is of shape ``chunks``,
+    except possibly for the last arrays along each dimension (if ``chunks``
+    do not even divide ``shape``).
+    """
+    ranges = [range(math.ceil(s / c)) for s, c in zip(shape, chunks)]
+    for indices in itertools.product(*ranges):
+        yield tuple(
+            slice(c * i, min(c * (i + 1), s)) for i, s, c in zip(indices, shape, chunks)
+        )
+
+
+def copy_stage(
+    source: ReadableArray, target: WriteableArray, chunks: Tuple[int, ...]
+) -> Stage:
+    # use a closure to eliminate extra arguments
+    def _copy_chunk(chunk_key):
+        # calling np.asarray here allows the source to be a dask array
+        # TODO: could we asyncify this to operate in a streaming fashion
+        # make sure this is not happening inside a dask scheduler
+        print(f"_copy_chunk({chunk_key})")
+        with dask.config.set(scheduler="single-threaded"):
+            data = np.asarray(source[chunk_key])
+        target[chunk_key] = data
+
+    keys = list(chunk_keys(source.shape, chunks))
+    return Stage(_copy_chunk, keys)
+
+
+def spec_to_pipeline(spec: CopySpec) -> MultiStagePipeline:
+    pipeline = []
+    if spec.intermediate.array is None:
+        pipeline.append(copy_stage(spec.read.array, spec.write.array, spec.read.chunks))
+    else:
+        pipeline.append(
+            copy_stage(spec.read.array, spec.intermediate.array, spec.read.chunks)
+        )
+        pipeline.append(
+            copy_stage(spec.intermediate.array, spec.write.array, spec.write.chunks)
+        )
+    return pipeline
+
+
+def specs_to_pipelines(specs: Iterable[CopySpec]) -> ParallelPipelines:
+    return [spec_to_pipeline(spec) for spec in specs]
+
+
+# TODO: understand how to make these type annotations work with a mixin
+T = TypeVar("T")
+
+
+class CopySpecToPipelinesMixin:
+    # This signature doesn't work as a mixin because we don't know what type T is
+    def prepare_plan(self, specs: Iterable[CopySpec]) -> T:
```

Comment:
@shoyer - I could use your advice on how to fix the type hints here. I am trying to use a mixin to transform `PipelineExecutor` to a `CopySpecExecutor`. The problem is that this mixin doesn't know what type `T`. This goes beyond my type hinting abilities...

Reply:
Another ping on @shoyer for a little help with type hints.

Reply:
I haven't tried to use type hints with mixins before. StackOverflow suggests a few approaches that may work: https://stackoverflow.com/questions/51930339/how-do-i-correctly-add-type-hints-to-mixin-classes